File: 2021/deno/day-7.ts
const input = await Deno.readTextFile("./input/day-7.in");

const positions = input.split(",").map(Number);

const sorted = positions.sort((a, b) => a - b);

const length = sorted.length;

const median =
  length % 2 === 0
    ? Math.floor((sorted[length / 2] + sorted[length / 2 - 1]) / 2)
    : sorted[Math.floor(length / 2)];

const total = sorted.reduce((prev, curr) => {
  return prev + Math.abs(curr - median);
}, 0);

/**
 * Part One
 */
console.log("Part One:", total);

/**
 * Part Two
 */

const maxPos = Math.max(...positions);

let superTotal = Infinity;
let current = sorted[0];

while (current <= maxPos) {
  const next = sorted.reduce((prev, x) => {
    const diff = Math.abs(x - current);
    const accumulated = (diff * (diff + 1)) / 2;

    return prev + accumulated;
  }, 0);

  if (next < superTotal) {
    superTotal = next;
  }

  current += 1;
}

console.log("Part Two:", superTotal);

File: 2021/swift/day-7.swift
import Foundation

let filename = "../input/day-7.in"

func main() {
    do {
        let contents = try String(contentsOfFile: filename)

        let positions = contents.trimmingCharacters(in: .whitespacesAndNewlines).split(separator: ",").map { Int($0)! }

        let length = positions.count

        let linear_optimal = {
            var sorted = positions
            sorted.sort()

            return (sorted[length / 2] + sorted[length / 2 + 1]) / 2
        }()

        let linear_total_fuel = positions.reduce(0) { $0 + abs($1 - linear_optimal) }

        print("Part one:", linear_total_fuel)

        let quad_fuel_calculator = {
            (steps: Int) -> Int in
            steps * (steps + 1) / 2
        }

        let quad_optimal = positions.reduce(0,+) / length

        let quad_total_fuel = positions.reduce(0) { $0 + quad_fuel_calculator(abs($1 - quad_optimal)) }

        print("Part two:", quad_total_fuel)

    } catch {
        print(error)
    }
}

main()
